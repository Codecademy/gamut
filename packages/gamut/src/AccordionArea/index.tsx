import styled from '@emotion/styled';
import { motion } from 'framer-motion';
import React, { useState } from 'react';
import { useIsomorphicLayoutEffect } from 'react-use';

export type AccordionAreaProps = {
  children: React.ReactNode;

  className?: string;

  /**
   * Whether the accordion is visually expanded to show its contents.
   */
  expanded?: boolean;

  /**
   * Contents of the clickable header button.
   */
  top: React.ReactNode;
};

const transitionDuration = 0.2;

const variants = {
  expanded: { height: 'auto' },
  folded: { height: '0' },
};

/**
 * @deprecated
 * This component is in the old visual identity and will be updated soon.
 *
 * Check the [Gamut Board](https://www.notion.so/codecademy/Gamut-Status-Timeline-dd3c135d3848464ea6eb1b48e68fbb1d) for component status
 */

export const AccordionArea: React.FC<AccordionAreaProps> = ({
  children,
  className,
  expanded,
  top,
}) => {
  const [delayExpanded, setDelayExpanded] = useState(expanded);

  useIsomorphicLayoutEffect(() => {
    const handle = setTimeout(
      () => setDelayExpanded(expanded),
      transitionDuration * 1000
    );

    return () => clearTimeout(handle);
  }, [expanded]);

  return (
    <div className={className}>
      {top}
      <StyledAccordionBody
        aria-expanded={expanded}
        initial={expanded ? 'expanded' : 'folded'}
        animate={expanded ? 'expanded' : 'folded'}
        variants={variants}
        transition={{ duration: transitionDuration, ease: 'easeInOut' }}
      >
        {(expanded || delayExpanded) && children}
      </StyledAccordionBody>
    </div>
  );
};

const StyledAccordionBody = styled(motion.div)`
  overflow: hidden;
`;
