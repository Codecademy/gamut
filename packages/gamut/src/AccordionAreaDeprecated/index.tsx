import styled from '@emotion/styled';
import { motion } from 'framer-motion';
import { useState } from 'react';
import * as React from 'react';
import { useIsomorphicLayoutEffect } from 'react-use';

export type AccordionAreaDeprecatedProps = {
  children: React.ReactNode;

  className?: string;

  /**
   * Whether the accordion is visually expanded to show its contents.
   */
  expanded?: boolean;

  /**
   * Contents of the clickable header button.
   */
  top: React.ReactNode;
};

const transitionDuration = 0.2;

const variants = {
  expanded: { height: 'auto' },
  folded: { height: '0' },
};

/**
 * @deprecated
 * This component is in the old visual identity and will be updated soon.
 *
 * Check the [Gamut Board](https://www.notion.so/codecademy/Gamut-Status-Timeline-dd3c135d3848464ea6eb1b48e68fbb1d) for component status
 */

export const AccordionAreaDeprecated: React.FC<
  AccordionAreaDeprecatedProps
> = ({ children, className, expanded, top }) => {
  const [delayExpanded, setDelayExpanded] = useState(expanded);

  useIsomorphicLayoutEffect(() => {
    const handle = setTimeout(
      () => setDelayExpanded(expanded),
      transitionDuration * 1000
    );

    return () => clearTimeout(handle);
  }, [expanded]);

  return (
    <div className={className}>
      {top}
      <StyledAccordionBody
        animate={expanded ? 'expanded' : 'folded'}
        initial={expanded ? 'expanded' : 'folded'}
        transition={{ duration: 0.2, ease: 'easeInOut' }}
        variants={variants}
      >
        {(expanded || delayExpanded) && children}
      </StyledAccordionBody>
    </div>
  );
};

const StyledAccordionBody = styled(motion.div)`
  overflow: hidden;
`;
