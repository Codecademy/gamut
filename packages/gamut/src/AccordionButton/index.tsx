import { ArrowChevronDownIcon } from '@codecademy/gamut-icons';
import cx from 'classnames';
import React from 'react';

import { ButtonDeprecated } from '../ButtonDeprecated';
import {
  ButtonDeprecatedBase,
  ButtonDeprecatedBaseProps,
} from '../ButtonDeprecatedBase';
import styles from './styles.module.scss';

export type AccordionButtonSize = 'normal' | 'large';

export type AccordionButtonTheme = 'blue' | 'plain' | 'yellow';

export type AccordionButtonProps = ButtonDeprecatedBaseProps & {
  /**
   * Whether the button should display as open or closed.
   */
  expanded?: boolean;

  /**
   * Determines the size of the button.
   */
  size?: 'normal' | 'large';

  /**
   * Visual theme for the clickable header button.
   */
  theme?: 'blue' | 'plain' | 'yellow';
};

const buttonThemes = {
  blue: {
    component: ButtonDeprecated,
    props: {
      flat: true,
      theme: 'white',
    },
  },
  plain: {
    component: ButtonDeprecated,
    props: {
      flat: true,
      theme: 'brand-dark-blue',
    },
  },
  yellow: {
    component: ButtonDeprecatedBase,
    props: {},
  },
} as const;

/**
 * @deprecated
 * This component is in the old visual identity and will be updated soon.
 *
 * Check the [Gamut Board](https://www.notion.so/codecademy/Gamut-Status-Timeline-dd3c135d3848464ea6eb1b48e68fbb1d) for component status
 */

export const AccordionButton: React.FC<AccordionButtonProps> = ({
  children,
  className,
  expanded,
  size = 'normal',
  theme = 'plain',
  ...baseProps
}) => {
  const { component: ButtonComponent, props } = buttonThemes[theme];
  const iconSize = size === 'large' ? 30 : undefined;

  return (
    <ButtonComponent
      aria-expanded={expanded}
      className={cx(
        styles.accordionButton,
        styles[theme],
        styles[size],
        className
      )}
      {...baseProps}
      {...props}
    >
      <span className={styles.children}>{children}</span>
      <ArrowChevronDownIcon
        className={cx(
          styles.expansionIcon,
          expanded && styles.expansionIconExpanded
        )}
        height={iconSize}
        width={iconSize}
        aria-hidden
      />
    </ButtonComponent>
  );
};
