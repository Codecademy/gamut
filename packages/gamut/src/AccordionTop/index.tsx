import cx from 'classnames';
import React from 'react';

import Button from '../Button';
import ChevronDownIcon from '../Icon/icons/ChevronDownIcon';
import styles from './styles.module.scss';
import ButtonBase from '../ButtonBase';

export type AccordionTopProps = {
  className?: string;
  expanded?: boolean;
  onClick: () => void;
  theme?: 'blue' | 'plain' | 'yellow';
};

const buttonThemes = {
  blue: {
    component: Button,
    props: {
      flat: true,
      theme: 'white',
    },
  },
  plain: {
    component: Button,
    props: {
      flat: true,
      outline: true,
      theme: 'brand-dark-blue',
    },
  },
  yellow: {
    component: ButtonBase,
    props: {},
  },
};

export const AccordionTop: React.FC<AccordionTopProps> = ({
  children,
  className,
  expanded,
  onClick,
  theme,
}) => {
  const { component: ButtonComponent, props } = buttonThemes[theme];

  return (
    <ButtonComponent
      aria-expanded={expanded}
      className={cx(styles.accordionTop, styles[theme], className)}
      onClick={onClick}
      {...props}
    >
      {children}
      <ChevronDownIcon
        className={cx(
          styles.expansionIcon,
          expanded && styles.expansionIconExpanded
        )}
      />
    </ButtonComponent>
  );
};

export default AccordionTop;
