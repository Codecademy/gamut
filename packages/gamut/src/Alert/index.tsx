export * from './Alert';
export * from './AlertArea';
