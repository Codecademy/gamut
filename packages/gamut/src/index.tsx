export * from './Anchor';
export * from './Alert';
export * from './Accordion';
export * from './AccordionArea';
export * from './AccordionButton';
export * from './AppBar';
export * from './AppBar/AppBarSection';
export * from './AppWrapper';
export * from './Badge';
export * from './Banner';
export * from './Box';
export * from './Button';
export * from './ButtonDeprecated';
export * from './ButtonDeprecatedBase';
export * from './BodyPortal';
export * from './Card';
export * from './DeprecatedCard/CardShell';
export * from './ContentContainer';
export * from './FlexBox';
export * from './FocusTrap';
export * from './Form';
export * from './GridForm';
export * from './HiddenText';
export * from './Typography/HeadingDeprecated';
export * from './Typography/Text';
export * from './Typography/TextDeprecated';
export * from './HighlightedText';
export * from './InputStepper';
export * from './Interstitial';
export * from './Layout/Column';
export * from './Layout/LayoutGrid';
export * from './Markdown';
export * from './Menus/MenuItem';
export * from './Menus/SideMenu';
export * from './Dialog';
export * from './ModalDeprecated';
export * from './NotificationList';
export * from './NotificationList/NotificationItem';
export * from './Overlay';
export * from './Pattern';
export * from './ProgressBar';
export * from './RadialProgress';
export * from './SkipToContent';
export * from './Spinner';
export * from './theming/VisualTheme';
export * from './Tabs';
export * from './Toggle';
export * from './ToolTip';
export * from './Truncate';
export * from './utils';
export * from './Video';
export * from './typings/responsive-properties';
export * from './Logo';
export * from './ProLabel';
export * from './Toast/Toast';
export * from './FloatingCard/FloatingCard';
