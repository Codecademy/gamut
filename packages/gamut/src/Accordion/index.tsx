import React, { useState } from 'react';

import { AccordionArea } from '../AccordionArea';
import {
  AccordionButton,
  AccordionButtonSize,
  AccordionButtonTheme,
} from '../AccordionButton';

export type ChildrenOrExpandedRender =
  | React.ReactNode
  | ((expanded: boolean) => React.ReactNode);

export type AccordionProps = {
  children: ChildrenOrExpandedRender;

  /**
   * CSS class name added to the root area container.
   */
  className?: string;

  /**
   * Whether the accordion should start off with expanded state.
   */
  initiallyExpanded?: boolean;

  /**
   * Called when the top button is clicked.
   *
   * @param expanding - New expanded state the accordion will transition to.
   */
  onClick?: (expanding: boolean) => void;

  /**
   * Visual size of the top button.
   */
  size?: AccordionButtonSize;

  /**
   * Visual theme of the top button.
   */
  theme?: AccordionButtonTheme;

  /**
   * Contents to place within the top button.
   */
  top: ChildrenOrExpandedRender;
};

/**
 * @deprecated
 * This component is in the old visual identity and will be updated soon.
 *
 * Check the [Gamut Board](https://www.notion.so/codecademy/Gamut-Status-Timeline-dd3c135d3848464ea6eb1b48e68fbb1d) for component status
 */

export const Accordion: React.FC<AccordionProps> = ({
  children,
  className,
  initiallyExpanded,
  onClick,
  size,
  theme,
  top,
}) => {
  const [expanded, setExpanded] = useState(!!initiallyExpanded);
  const expandRenderer = (renderer: ChildrenOrExpandedRender) =>
    renderer instanceof Function ? renderer(expanded) : renderer;

  return (
    <AccordionArea
      className={className}
      expanded={expanded}
      top={
        <AccordionButton
          expanded={expanded}
          onClick={() => {
            setExpanded(!expanded);
            onClick?.(!expanded);
          }}
          size={size}
          theme={theme}
        >
          {expandRenderer(top)}
        </AccordionButton>
      }
    >
      {expandRenderer(children)}
    </AccordionArea>
  );
};
