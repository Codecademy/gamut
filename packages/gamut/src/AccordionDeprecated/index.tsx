import { useState } from 'react';
import * as React from 'react';

import { AccordionAreaDeprecated } from '../AccordionAreaDeprecated';
import {
  AccordionButtonDeprecated,
  AccordionButtonSize,
  AccordionButtonTheme,
} from '../AccordionButtonDeprecated';

export type ChildrenOrExpandedRender =
  | React.ReactNode
  | ((expanded: boolean) => React.ReactNode);

export type AccordionDeprecatedProps = {
  children: ChildrenOrExpandedRender;

  /**
   * CSS class name added to the root area container.
   */
  className?: string;

  /**
   * Whether the accordion should start off with expanded state.
   */
  initiallyExpanded?: boolean;

  /**
   * Called when the top button is clicked.
   *
   * @param expanding - New expanded state the accordion will transition to.
   */
  onClick?: (expanding: boolean) => void;

  /**
   * Visual size of the top button.
   */
  size?: AccordionButtonSize;

  /**
   * Visual theme of the top button.
   */
  theme?: AccordionButtonTheme;

  /**
   * Contents to place within the top button.
   */
  top: ChildrenOrExpandedRender;
};

/**
 * @deprecated
 * This component is in the old visual identity and will be updated soon.
 *
 * Check the [Gamut Board](https://www.notion.so/codecademy/Gamut-Status-Timeline-dd3c135d3848464ea6eb1b48e68fbb1d) for component status
 */

export const AccordionDeprecated: React.FC<AccordionDeprecatedProps> = ({
  children,
  className,
  initiallyExpanded,
  onClick,
  size,
  theme,
  top,
}) => {
  const [expanded, setExpanded] = useState(!!initiallyExpanded);
  const expandRenderer = (renderer: ChildrenOrExpandedRender) =>
    renderer instanceof Function ? renderer(expanded) : renderer;

  return (
    <AccordionAreaDeprecated
      className={className}
      expanded={expanded}
      top={
        <AccordionButtonDeprecated
          expanded={expanded}
          size={size}
          theme={theme}
          onClick={() => {
            setExpanded(!expanded);
            onClick?.(!expanded);
          }}
        >
          {expandRenderer(top)}
        </AccordionButtonDeprecated>
      }
    >
      {expandRenderer(children)}
    </AccordionAreaDeprecated>
  );
};
