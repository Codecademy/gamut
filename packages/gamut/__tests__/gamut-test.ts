import * as gamut from '../src/index';

test('Gamut Exported Keys', () => {
  const exportedNames = Object.keys(gamut).sort((a, b) =>
    a.toLowerCase().localeCompare(b.toLowerCase())
  );
  expect(exportedNames).toMatchSnapshot();
});
