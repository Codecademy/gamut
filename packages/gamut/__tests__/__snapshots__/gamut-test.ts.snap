// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gamut Exported Keys 1`] = `
Array [
  "Accordion",
  "AccordionButton",
  "Alert",
  "AppBar",
  "AppBarSection",
  "Badge",
  "Banner",
  "BannerStyle",
  "BannerType",
  "Button",
  "ButtonBase",
  "buttonPresetThemes",
  "CardBody",
  "CardFooter",
  "CardShell",
  "Checkbox",
  "Col",
  "Column",
  "Container",
  "ContentContainer",
  "createPromise",
  "DeprecatedTabs",
  "FlexGrid",
  "Form",
  "FormError",
  "FormGroup",
  "FormGroupDescription",
  "FormGroupLabel",
  "generateResponsiveClassnames",
  "GridForm",
  "Heading",
  "HighlightedText",
  "Input",
  "InputStepper",
  "Interstitial",
  "Item",
  "LayoutGrid",
  "Markdown",
  "MenuItem",
  "Modal",
  "NotificationIcon",
  "NotificationItem",
  "NotificationList",
  "omitProps",
  "Overlay",
  "ProgressBar",
  "RadialProgress",
  "Radio",
  "RadioGroup",
  "Row",
  "Select",
  "SideMenu",
  "SkipToContent",
  "Spinner",
  "SplitInterstitial",
  "Tab",
  "TabList",
  "TabPanel",
  "Tabs",
  "Text",
  "TextArea",
  "Toggle",
  "ToolTip",
  "ToolTipPosition",
  "Truncate",
  "Video",
  "VisualTheme",
]
`;
